cdcsdk-testing/src/test/java/com/yugabyte/cdcsdk/testing/PubSubConsumerIT.java

```diff
@@ -0,0 +1,126 @@
+        initializeContainers();
+
+        initHelpers(true, false, false);
+    }
+
+    @BeforeEach
+    public void beforeEachTest() throws Exception {
+        ybHelper.execute(UtilStrings.getCreateTableYBStmt(DEFAULT_TABLE_NAME));
+    }
+
+    @AfterEach
+    public void dropTable() throws Exception {
+        ybHelper.execute(UtilStrings.getDropTableStmt(DEFAULT_TABLE_NAME));
+    }
+
+    @Test
+    public void automationOfPubSubAssertions() throws Exception {
+
+        cdcsdkContainer = TestHelper
+                .getCdcsdkContainerForPubSubSink(ybHelper, projectId, "public." + DEFAULT_TABLE_NAME)
+                .withCreateContainerCmdModifier(cmd -> cmd.withUser(System.getenv("USERID")));
+        cdcsdkContainer.withNetwork(containerNetwork);
+
+        cdcsdkContainer.withFileSystemBind(System.getenv("GOOGLE_APPLICATION_CREDENTIALS"),
+                "/tmp/keys/application_default_credentials.json", BindMode.READ_ONLY);
+        cdcsdkContainer.addEnv("GOOGLE_APPLICATION_CREDENTIALS", "/tmp/keys/application_default_credentials.json");
+        cdcsdkContainer.start();
+
+        assertTrue(cdcsdkContainer.isRunning());
+
+        int recordsInserted = 1;
+        for (int i = 0; i < recordsInserted; ++i) {
+            ybHelper.execute(UtilStrings.getInsertStmt(DEFAULT_TABLE_NAME, i, "first_" + i, "last_" + i, 23.45));
+            Awaitility.await().atMost(Duration.ofSeconds(10));
+        }
+        ybHelper.execute(UtilStrings.getUpdateStmt(DEFAULT_TABLE_NAME, 0, 25));
+        Awaitility.await().atMost(Duration.ofSeconds(10));
+
+        ybHelper.execute(UtilStrings.getDeleteStmt(DEFAULT_TABLE_NAME, 0));
+        Awaitility.await().atMost(Duration.ofSeconds(10));
+
+        List<String> expected_data = List.of(
+                "{\"id\":0,\"first_name\":\"first_0\",\"last_name\":\"last_0\",\"days_worked\":23.45,\"__deleted\":\"false\"}",
+                "{\"id\":0,\"days_worked\":25.0,\"__deleted\":\"false\"}",
+                "{\"id\":0,\"__deleted\":\"true\"}");
+        Iterator<String> expected = expected_data.iterator();
+
+        List<String> actual_data = new ArrayList<>();
+
+        ProjectSubscriptionName subscriptionName = ProjectSubscriptionName.of(projectId, subscriptionId);
+
+        // Instantiate an asynchronous message receiver.
+        MessageReceiver receiver = (PubsubMessage message, AckReplyConsumer consumer) -> {
+            // Handle incoming message, then ack the received message.
+            actual_data.add(message.getData().toStringUtf8());
+            consumer.ack();
+        };
+
+        Subscriber subscriber = null;
+        try {
+            subscriber = Subscriber.newBuilder(subscriptionName, receiver).build();
+            // Start the subscriber.
+            subscriber.startAsync().awaitRunning();
+            // Allow the subscriber to run for 30s unless an unrecoverable error occurs.
+            subscriber.awaitTerminated(30, TimeUnit.SECONDS);
```

Comment:
Does the subscriber have a conditional wait? Is it possible to add a condition to wait until `actual_data` has 3 rows?

Reply:
Subscriber only has two methods to check if the API service has reached terminal state (One with timeout and the other one without timeout). Internally `awaitTerminated() ` checks the count of messages in pub/sub that are yet to be acked or nacked. Once the count reaches zero it moves to terminated state.